src/mlpack/tests/rbm_network_test.cpp

```diff
@@ -0,0 +1,210 @@
+#include <mlpack/core.hpp>
+
+#include <mlpack/core/optimizers/cdk/cdk.hpp>
+#include <mlpack/core/optimizers/rmsprop/rmsprop.hpp>
+#include <mlpack/methods/ann/init_rules/gaussian_init.hpp>
+#include <mlpack/methods/ann/layer/layer.hpp>
+#include <mlpack/methods/ann/rbm.hpp>
+#include <mlpack/methods/softmax_regression/softmax_regression.hpp>
+#include <mlpack/core/optimizers/sgd/sgd.hpp>
+
+#include <boost/test/unit_test.hpp>
+#include "test_tools.hpp"
+
+using namespace mlpack;
+using namespace mlpack::ann;
+using namespace mlpack::optimization;
+using namespace mlpack::regression;
+
+BOOST_AUTO_TEST_SUITE(RbmNetworkTest);
+
+template<typename MatType = arma::mat>
+void BuildVanillaNetwork(MatType& trainData,
+                         const size_t hiddenLayerSize)
+{
+  /*
+   * Construct a feed forward network with trainData.n_rows input nodes,
+   * hiddenLayerSize hidden nodes and trainLabels.n_rows output nodes. The
+   * network structure looks like:
+   *
+   *  Visible       Hidden        
+   *  Layer         Layer         
+   * +-----+       +-----+       
+   * |     |       |     |            
+   * |     +<----->|     |
+   * |     |       |     | 
+   * +-----+       +-----+     
+   *        
+   */
+  arma::mat output;
+  RandomSeed(23);
+  BinaryLayer<> visible(trainData.n_rows, hiddenLayerSize, 1);
+  BinaryLayer<> hidden(hiddenLayerSize, trainData.n_rows, 0);
+  GaussianInitialization gaussian(0, 0.1);
+  RBM<GaussianInitialization, BinaryLayer<>, BinaryLayer<> > model(trainData,
+      gaussian, visible, hidden, 1,  true);
+  CDK<RBM<GaussianInitialization, BinaryLayer<>, BinaryLayer<> >> cdk(model,
+      0.06, trainData.n_cols * 20, 10, true);
+  model.Reset();
+  // Set the parmaeters from a learned rbm
+  model.Parameters() = {-0.23224054, -0.23000632,
+                        -0.25701271, -0.25122418,
+                        -0.20716651,  -0.20962217,
+                        -0.59922456, -0.60003836,
+                        -0.6, -0.625, -0.475};
+  // Check Weight Shared
+  BOOST_REQUIRE_EQUAL(arma::accu(model.VisibleLayer().Weight() - 
+      model.HiddenLayer().Weight()), 0);
+
+  // Check free energy
+  arma::vec freeEnergy = {-0.87523715, 0.50615066, 0.46923476, 1.21509084};
+  arma::vec calcultedFreeEnergy(4);
+  calcultedFreeEnergy.zeros();
+  for (size_t i = 0; i < trainData.n_cols; i++)
+  {
+    model.VisibleLayer().ForwardPreActivation(std::move(trainData.col(i)),
+        std::move(output));
+    calcultedFreeEnergy(i) = model.FreeEnergy(std::move(trainData.col(i)));
+  }
+
+  BOOST_REQUIRE_LE(arma::accu(calcultedFreeEnergy - freeEnergy), 1);
+  
+}
+BOOST_AUTO_TEST_CASE(MiscTest)
+{
+  /**
+   * Train and evaluate a vanilla network with the specified structure.
+   */
+
+  arma::mat X;
```

Comment:
I will change the seed. The test actually does not depend upon the seed value if we intialize the parameter value.